### PR TITLE
Issue LIBCLOUD-365: change ec2 ex_blockdevicemappings to support all possible key/values

### PR DESCRIPTION
**Pull request by joemiller**

### CODE REVIEW

File: libcloud/compute/drivers/ec2.py [hunk 1]
```diff
@@ -1331,11 +1332,10 @@ def create_node(self, **kwargs):
...
 
         if 'ex_blockdevicemappings' in kwargs:
-            for index, mapping in enumerate(kwargs['ex_blockdevicemappings']):
-                params['BlockDeviceMapping.%d.DeviceName' % (index + 1)] = \
-                    mapping['DeviceName']
-                params['BlockDeviceMapping.%d.VirtualName' % (index + 1)] = \
-                    mapping['VirtualName']
+            for idx, mapping in enumerate(kwargs['ex_blockdevicemappings'],
+                                          start=1):
+                 for k, v in mapping.iteritems():
```
Comment (Kami) [inline libcloud/compute/drivers/ec2.py]: Please use `dict.items`. `dict.iteritems` has been removed in Python 3 (`dict.items` behaves pretty much the same as `dict.iteritems` in Python 2.x).

File: libcloud/compute/drivers/ec2.py [hunk 2]
```diff
@@ -1331,11 +1332,10 @@ def create_node(self, **kwargs):
             params['ClientToken'] = kwargs['ex_clienttoken']
 
         if 'ex_blockdevicemappings' in kwargs:
```
Comment (Kami) [inline libcloud/compute/drivers/ec2.py:1334]: I would also verify that `kwargs['ex_blockdevicemappings'] is a list / tuple.

File: libcloud/compute/drivers/ec2.py [hunk 3]
```diff
@@ -1331,11 +1332,10 @@ def create_node(self, **kwargs):
...
-            for index, mapping in enumerate(kwargs['ex_blockdevicemappings']):
-                params['BlockDeviceMapping.%d.DeviceName' % (index + 1)] = \
-                    mapping['DeviceName']
-                params['BlockDeviceMapping.%d.VirtualName' % (index + 1)] = \
-                    mapping['VirtualName']
+            for idx, mapping in enumerate(kwargs['ex_blockdevicemappings'],
```
Comment (Kami) [inline libcloud/compute/drivers/ec2.py:1338]: Same here, I would verify `mapping` is a dict.

**Comment by joemiller** [comment]
updated with type checking

File: libcloud/compute/drivers/ec2.py [hunk 4]
```diff
@@ -1331,11 +1332,17 @@ def create_node(self, **kwargs):
...
-                    mapping['DeviceName']
-                params['BlockDeviceMapping.%d.VirtualName' % (index + 1)] = \
-                    mapping['VirtualName']
+            if not isinstance(kwargs['ex_blockdevicemappings'], list):
+                raise AttributeError('ex_blockdevicemappings is not a list')
+
+            for idx, mapping in enumerate(kwargs['ex_blockdevicemappings'],
+                                          start=1):
+                for k, v in mapping.items():
+                    if not isinstance(mapping, dict):
```
Comment (Kami) [inline libcloud/compute/drivers/ec2.py:1341]: Please put this outside this inner for loop.

File: libcloud/compute/drivers/ec2.py [hunk 5]
```diff
@@ -1331,11 +1332,17 @@ def create_node(self, **kwargs):
             params['ClientToken'] = kwargs['ex_clienttoken']
 
         if 'ex_blockdevicemappings' in kwargs:
-            for index, mapping in enumerate(kwargs['ex_blockdevicemappings']):
-                params['BlockDeviceMapping.%d.DeviceName' % (index + 1)] = \
-                    mapping['DeviceName']
-                params['BlockDeviceMapping.%d.VirtualName' % (index + 1)] = \
-                    mapping['VirtualName']
+            if not isinstance(kwargs['ex_blockdevicemappings'], list):
```
Comment (Kami) [inline libcloud/compute/drivers/ec2.py:1335]: I would do `not isinstance(kwargs['ex_blockdevicemappings'], (list, tuple))`

**Comment by Kami** [comment]
@joemiller Thanks.

When you address those two last comments, please squash the commits, attach patch to the JIRA ticket and I'll merge it into trunk.

**Comment by Kami** [comment]
@joemiller bump - how are those changes looking?

**Comment by joemiller** [comment]
Sorry, been very busy this week. I will do my best to get to them tonight.

**Comment by joemiller** [comment]
patch attached : https://issues.apache.org/jira/browse/LIBCLOUD-365

**Comment by Kami** [comment]
@joemiller Thanks, i've merged your pr into trunk and 0.13.x.

**Comment by joemiller** [comment]
Tomaz, great, thank you!

Did you mean 0.13.x branch btw?

**Comment by Kami** [comment]
@joemiller Yeah, sorry, I've meant 0.13.x (I've also updated my comment) :)

**Comment by Kami** [comment]
Just FYI - this pr can be closed now. Thanks again!

**Pull request closed without merging**